File: app/app_peripherals_bridge.cpp
/**
 * @file app_peripherals_bridge.cpp
 * @brief 外设统一 SDK bridge 应用
 *
 * 本应用在一个进程内只创建一个机器人 SDK 会话，然后按需启动
 * SpaceMouse 与/或六维力/力矩传感器 reader，并通过
 * `Robot::Peripheral().UpdateSpaceMouseSample()` /
 * `UpdateFtSensorSample()` 将采样送入机器人控制器。
 *
 * 使用方式：
 * - FDCC 同时需要 SpaceMouse 与力传感器：
 *   `app_peripherals_bridge --robot <robot_ip>`
 * - 只启用 SpaceMouse：
 *   `app_peripherals_bridge --robot <robot_ip> --spacemouse`
 * - 只启用力传感器：
 *   `app_peripherals_bridge --robot <robot_ip> --ft-sensor`
 * - 同时显式启用两者：
 *   `app_peripherals_bridge --robot <robot_ip> --spacemouse --ft-sensor`
 *
 * 选择规则：不传 `--spacemouse`/`--ft-sensor` 时默认两路都启动；只要出现
 * 任一选择参数，则仅启动显式选择的外设。设备参数只影响对应外设，例如
 * `--spacemouse-device /dev/input/event7`、`--ft-serial-port /dev/ttyUSB0`。
 *
 * 构建要求：仅在 `SMR_PERIPHERAL_WITH_SDK=ON`（`--with-sdk ON`）时编译本目标。
 */

#include "peripherals/peripherals.hpp"
#include "sdk/robot.hpp"

#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <mutex>
#include <string>
#include <thread>

namespace
{

std::atomic<bool> g_running{true};

void HandleSignal(int)
{
    g_running.store(false, std::memory_order_release);
}

struct BridgeOptions
{
    std::string robot_ip;
    bool spacemouse_enabled{true};
    bool ft_sensor_enabled{true};
    bool help_requested{false};
    smrcore::peripherals::SpaceMouseOptions spacemouse;
    smrcore::peripherals::FtSensorOptions ft_sensor;
};

void PrintUsage(const char *program)
{
    std::cerr
        << "usage: " << program << " [--robot IP] "
        << "[--spacemouse] [--ft-sensor]\n"
        << "       [--spacemouse-device PATH] "
           "[--spacemouse-sample-rate HZ]\n"
        << "       [--ft-serial-port PATH] [--ft-sensor-type TYPE] "
           "[--ft-baud-rate BAUD]\n"
        << "       [--ft-sample-rate HZ] [--ft-stale-timeout-ms MS]\n";
}

bool ParseCli(int argc, char **argv, BridgeOptions &options)
{
    bool explicit_enable = false;
    options.ft_sensor.serial_port = "/dev/ttyUSB0";
    options.ft_sensor.sample_rate_hz = 1000.0;

    for (int i = 1; i < argc; ++i)
    {
        const std::string arg = argv[i];
        if (arg == "--robot" && i + 1 < argc)
        {
            options.robot_ip = argv[++i];
        }
        else if (arg == "--spacemouse")
        {
            if (!explicit_enable)
            {
                options.spacemouse_enabled = false;
                options.ft_sensor_enabled = false;
                explicit_enable = true;
            }
            options.spacemouse_enabled = true;
        }
        else if (arg == "--ft-sensor")
        {
            if (!explicit_enable)
            {
                options.spacemouse_enabled = false;
                options.ft_sensor_enabled = false;
                explicit_enable = true;
            }
            options.ft_sensor_enabled = true;
        }
        else if (arg == "--spacemouse-device" && i + 1 < argc)
        {
            options.spacemouse.device_path = argv[++i];
        }
        else if (arg == "--spacemouse-sample-rate" && i + 1 < argc)
        {
            options.spacemouse.sample_rate_hz = std::atof(argv[++i]);
        }
        else if (arg == "--ft-serial-port" && i + 1 < argc)
        {
            options.ft_sensor.serial_port = argv[++i];
        }
        else if (arg == "--ft-sensor-type" && i + 1 < argc)
        {
            options.ft_sensor.sensor_type = argv[++i];
        }
        else if (arg == "--ft-baud-rate" && i + 1 < argc)
        {
            options.ft_sensor.baud_rate = std::atoi(argv[++i]);
        }
        else if (arg == "--ft-sample-rate" && i + 1 < argc)
        {
            options.ft_sensor.sample_rate_hz = std::atof(argv[++i]);
        }
        else if (arg == "--ft-stale-timeout-ms" && i + 1 < argc)
        {
            options.ft_sensor.stale_timeout_ms = std::atoi(argv[++i]);
        }
        else if (arg == "--help" || arg == "-h")
        {
            PrintUsage(argv[0]);
            options.help_requested = true;
            return true;
        }
        else
        {
            std::cerr << "unknown or incomplete argument: " << arg << '\n';
            PrintUsage(argv[0]);
            return false;
        }
    }

    return true;
}

rcore::GripperCommand ToSdk(smrcore::peripherals::GripperCommand command)
{
    return command == smrcore::peripherals::GripperCommand::Close
               ? rcore::GripperCommand::Close
               : rcore::GripperCommand::Open;
}

rcore::SpaceMouseSample ToSdk(
    const smrcore::peripherals::SpaceMouseSample &sample)
{
    rcore::SpaceMouseSample out;
    out.x = sample.x;
    out.y = sample.y;
    out.z = sample.z;
    out.roll = sample.roll;
    out.pitch = sample.pitch;
    out.yaw = sample.yaw;
    out.gripper_command = ToSdk(sample.gripper_command);
    out.timestamp_sec = sample.timestamp_sec;
    return out;
}

rcore::FtSensorSample ToSdk(const smrcore::peripherals::FtSensorSample &sample)
{
    rcore::FtSensorSample out;
    out.fx = sample.fx;
    out.fy = sample.fy;
    out.fz = sample.fz;
    out.tx = sample.tx;
    out.ty = sample.ty;
    out.tz = sample.tz;
    out.timestamp_sec = sample.timestamp_sec;
    return out;
}

void RunSpaceMouseBridge(smrcore::peripherals::SpaceMouse &spacemouse,
                         rcore::sdk::PeripheralApi &peripheral,
                         std::mutex &sdk_mutex)
{
    while (g_running.load(std::memory_order_acquire))
    {
        if (auto sample = spacemouse.GetSample())
        {
            std::lock_guard<std::mutex> lock(sdk_mutex);
            auto result = peripheral.UpdateSpaceMouseSample(ToSdk(*sample));
            if (!result.IsSuccess())
            {
                std::cerr << "UpdateSpaceMouseSample failed: "
                          << result.GetErrorCode() << " "
                          << result.GetErrorMsg() << std::endl;
            }
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(8));
    }
}

void RunFtSensorBridge(smrcore::peripherals::FtSensor &sensor,
                       rcore::sdk::PeripheralApi &peripheral,
                       std::mutex &sdk_mutex)
{
    while (g_running.load(std::memory_order_acquire))
    {
        if (auto sample = sensor.GetSample())
        {
            std::lock_guard<std::mutex> lock(sdk_mutex);
            auto result = peripheral.UpdateFtSensorSample(ToSdk(*sample));
            if (!result.IsSuccess())
            {
                std::cerr << "UpdateFtSensorSample failed: "
                          << result.GetErrorCode() << " "
                          << result.GetErrorMsg() << std::endl;
            }
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
}

} // namespace

int main(int argc, char **argv)
{
    BridgeOptions options;
    if (!ParseCli(argc, argv, options))
    {
        return 1;
    }
    if (options.help_requested)
    {
        return 0;
    }

    std::signal(SIGINT, HandleSignal);
    std::signal(SIGTERM, HandleSignal);

    rcore::sdk::Robot robot;
    std::cout << "正在连接机器人"
              << (options.robot_ip.empty() ? "" : " " + options.robot_ip)
              << " ..." << std::endl;
    if (!robot.Initialize(options.robot_ip))
    {
        std::cerr << "failed to initialize robot SDK" << std::endl;
        return 1;
    }
    std::cout << "机器人 SDK 连接成功" << std::endl;

    smrcore::peripherals::SpaceMouse spacemouse;
    if (options.spacemouse_enabled)
    {
        std::cout << "正在启动 SpaceMouse ..." << std::endl;
        if (!spacemouse.Initialize(options.spacemouse) || !spacemouse.Start())
        {
            std::cerr << "failed to start SpaceMouse reader" << std::endl;
            return 1;
        }
        std::cout << "SpaceMouse 已启动" << std::endl;
    }

    smrcore::peripherals::FtSensor sensor;
    if (options.ft_sensor_enabled)
    {
        std::cout << "正在启动力传感器 " << options.ft_sensor.serial_port
                  << " ..." << std::endl;
        if (!sensor.Initialize(options.ft_sensor) || !sensor.Start())
        {
            std::cerr << "failed to start F/T sensor reader" << std::endl;
            return 1;
        }
        if (!sensor.WaitForFirstSample())
        {
            const auto info = sensor.GetInfo();
            std::cerr << "first valid F/T frame timeout, frame_errors="
                      << info.frame_error_count << std::endl;
            return 1;
        }
        std::cout << "力传感器已启动" << std::endl;
    }

    auto peripheral = robot.Peripheral();
    std::mutex sdk_mutex;
    std::thread spacemouse_thread;
    std::thread ft_sensor_thread;

    if (options.spacemouse_enabled)
    {
        spacemouse_thread = std::thread(RunSpaceMouseBridge,
                                        std::ref(spacemouse),
                                        std::ref(peripheral),
                                        std::ref(sdk_mutex));
    }
    if (options.ft_sensor_enabled)
    {
        ft_sensor_thread = std::thread(RunFtSensorBridge, std::ref(sensor),
                                       std::ref(peripheral),
                                       std::ref(sdk_mutex));
    }

    std::cout << "连接成功，正在发送外设消息"
              << (options.spacemouse_enabled ? " [SpaceMouse]" : "")
              << (options.ft_sensor_enabled ? " [F/T]" : "")
              << "（Ctrl+C 退出）" << std::endl;

    if (spacemouse_thread.joinable())
    {
        spacemouse_thread.join();
    }
    if (ft_sensor_thread.joinable())
    {
        ft_sensor_thread.join();
    }
    std::cout << "外设 bridge 已停止" << std::endl;
    return 0;
}

File: python/app/app_peripherals_bridge.py
#!/usr/bin/env python3
"""Mirror of app/app_peripherals_bridge.cpp.

单一机器人 SDK 会话内按 CLI 启动 SpaceMouse 和/或 F/T reader，并通过
`robot.Peripheral().Update*Sample()` 将采样送入机器人控制器。

依赖（需分别安装 wheel）：
  pip install rcore-peripherals-py rcore-sdk-py
"""

from __future__ import annotations

import argparse
import signal
import sys
import threading
import time
from dataclasses import dataclass

try:
    from rcore_sdk import (
        FtSensorSample as SdkFtSensorSample,
        Robot,
        SpaceMouseSample as SdkSpaceMouseSample,
    )
except ImportError as exc:
    raise SystemExit(
        "rcore_sdk is required for the bridge app. Install the Python robot SDK wheel."
    ) from exc

from rcore_peripherals import FtSensor, FtSensorOptions, SpaceMouse, SpaceMouseOptions

_running = True
_running_lock = threading.Lock()


def _set_running(value: bool) -> None:
    global _running
    with _running_lock:
        _running = value


def _is_running() -> bool:
    with _running_lock:
        return _running


def _handle_signal(*_args: object) -> None:
    _set_running(False)


def _to_sdk_spacemouse(sample: object) -> SdkSpaceMouseSample:
    return SdkSpaceMouseSample(
        x=sample.x,
        y=sample.y,
        z=sample.z,
        roll=sample.roll,
        pitch=sample.pitch,
        yaw=sample.yaw,
        gripper_command=sample.gripper_command,
        timestamp_sec=sample.timestamp_sec,
    )


def _to_sdk_ft_sensor(sample: object) -> SdkFtSensorSample:
    return SdkFtSensorSample(
        fx=sample.fx,
        fy=sample.fy,
        fz=sample.fz,
        tx=sample.tx,
        ty=sample.ty,
        tz=sample.tz,
        timestamp_sec=sample.timestamp_sec,
    )


def _run_spacemouse_bridge(
    spacemouse: SpaceMouse,
    peripheral: object,
    sdk_lock: threading.Lock,
) -> None:
    while _is_running():
        sample = spacemouse.GetSample()
        if sample is not None:
            with sdk_lock:
                result = peripheral.UpdateSpaceMouseSample(_to_sdk_spacemouse(sample))
            if not result.success:
                print(
                    "UpdateSpaceMouseSample failed:",
                    result.error_code,
                    result.error_msg,
                    file=sys.stderr,
                    flush=True,
                )
        time.sleep(0.008)


def _run_ft_sensor_bridge(
    sensor: FtSensor,
    peripheral: object,
    sdk_lock: threading.Lock,
) -> None:
    while _is_running():
        sample = sensor.GetSample()
        if sample is not None:
            with sdk_lock:
                result = peripheral.UpdateFtSensorSample(_to_sdk_ft_sensor(sample))
            if not result.success:
                print(
                    "UpdateFtSensorSample failed:",
                    result.error_code,
                    result.error_msg,
                    file=sys.stderr,
                    flush=True,
                )
        time.sleep(0.001)


@dataclass
class BridgeOptions:
    robot_ip: str = ""
    spacemouse_enabled: bool = True
    ft_sensor_enabled: bool = True
    spacemouse: SpaceMouseOptions | None = None
    ft_sensor: FtSensorOptions | None = None

    def __post_init__(self) -> None:
        if self.spacemouse is None:
            self.spacemouse = SpaceMouseOptions()
        if self.ft_sensor is None:
            self.ft_sensor = FtSensorOptions(
                serial_port="/dev/ttyUSB0",
                sample_rate_hz=1000.0,
            )


def parse_args(argv: list[str] | None = None) -> BridgeOptions:
    parser = argparse.ArgumentParser(
        description="Bridge SpaceMouse and/or F/T sensor samples into robot SDK"
    )
    parser.add_argument("--robot-ip", default="", help="Robot controller IP (optional)")
    parser.add_argument("--spacemouse", action="store_true", help="Enable SpaceMouse")
    parser.add_argument("--ft-sensor", action="store_true", help="Enable F/T sensor")
    parser.add_argument("--spacemouse-device", default="", help="SpaceMouse device path")
    parser.add_argument(
        "--spacemouse-sample-rate",
        type=float,
        default=0.0,
        help="SpaceMouse sample rate [Hz]",
    )
    parser.add_argument("--ft-serial-port", default="/dev/ttyUSB0")
    parser.add_argument("--ft-sensor-type", default="xjc_serial")
    parser.add_argument("--ft-baud-rate", type=int, default=460800)
    parser.add_argument("--ft-sample-rate", type=float, default=1000.0)
    parser.add_argument("--ft-stale-timeout-ms", type=int, default=20)
    args = parser.parse_args(argv)

    options = BridgeOptions(
        robot_ip=args.robot_ip,
        spacemouse=SpaceMouseOptions(
            device_path=args.spacemouse_device,
            sample_rate_hz=args.spacemouse_sample_rate,
        ),
        ft_sensor=FtSensorOptions(
            serial_port=args.ft_serial_port,
            sensor_type=args.ft_sensor_type,
            baud_rate=args.ft_baud_rate,
            sample_rate_hz=args.ft_sample_rate,
            stale_timeout_ms=args.ft_stale_timeout_ms,
        ),
    )
    if args.spacemouse or args.ft_sensor:
        options.spacemouse_enabled = bool(args.spacemouse)
        options.ft_sensor_enabled = bool(args.ft_sensor)
    return options


def main(argv: list[str] | None = None) -> int:
    options = parse_args(argv)

    signal.signal(signal.SIGINT, _handle_signal)
    signal.signal(signal.SIGTERM, _handle_signal)

    robot = Robot()
    ip_suffix = f" {options.robot_ip}" if options.robot_ip else ""
    print(f"正在连接机器人{ip_suffix} ...", flush=True)
    if not robot.Initialize(options.robot_ip):
        print("failed to initialize robot SDK", file=sys.stderr, flush=True)
        return 1
    print("机器人 SDK 连接成功", flush=True)

    spacemouse = SpaceMouse()
    if options.spacemouse_enabled:
        assert options.spacemouse is not None
        print("正在启动 SpaceMouse ...", flush=True)
        if not spacemouse.Initialize(options.spacemouse) or not spacemouse.Start():
            print("failed to start SpaceMouse reader", file=sys.stderr, flush=True)
            return 1
        print("SpaceMouse 已启动", flush=True)

    sensor = FtSensor()
    if options.ft_sensor_enabled:
        assert options.ft_sensor is not None
        print(f"正在启动力传感器 {options.ft_sensor.serial_port} ...", flush=True)
        if not sensor.Initialize(options.ft_sensor) or not sensor.Start():
            print("failed to start F/T sensor reader", file=sys.stderr, flush=True)
            return 1
        if not sensor.WaitForFirstSample():
            info = sensor.GetInfo()
            print(
                "first valid F/T frame timeout, frame_errors="
                f"{info.frame_error_count}",
                file=sys.stderr,
                flush=True,
            )
            return 1
        print("力传感器已启动", flush=True)

    peripheral = robot.Peripheral()
    sdk_lock = threading.Lock()
    threads: list[threading.Thread] = []

    if options.spacemouse_enabled:
        thread = threading.Thread(
            target=_run_spacemouse_bridge,
            args=(spacemouse, peripheral, sdk_lock),
            name="spacemouse-bridge",
        )
        thread.start()
        threads.append(thread)

    if options.ft_sensor_enabled:
        thread = threading.Thread(
            target=_run_ft_sensor_bridge,
            args=(sensor, peripheral, sdk_lock),
            name="ft-sensor-bridge",
        )
        thread.start()
        threads.append(thread)

    channels = ""
    if options.spacemouse_enabled:
        channels += " [SpaceMouse]"
    if options.ft_sensor_enabled:
        channels += " [F/T]"
    print(f"连接成功，正在发送外设消息{channels}（Ctrl+C 退出）", flush=True)

    try:
        while _is_running():
            time.sleep(0.1)
    finally:
        _set_running(False)
        for thread in threads:
            thread.join()
        robot.Shutdown()
        print("外设 bridge 已停止", flush=True)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
